openpype/hosts/fusion/plugins/publish/validate_saver_resolution.py

```diff
@@ -87,19 +38,80 @@ def process(self, instance):
+        saver = instance.data["tool"]
+        try:
+            resolution = cls.get_resolution(instance)
+        except PublishValidationError:
+            resolution = None
         expected_resolution = cls.get_expected_resolution(instance)
         if resolution != expected_resolution:
-            saver = instance.data["tool"]
             return [saver]
 
     @classmethod
     def get_resolution(cls, instance):
         saver = instance.data["tool"]
         first_frame = instance.data["frameStartHandle"]
-        return get_tool_resolution(saver, frame=first_frame)
+        return cls.get_tool_resolution(saver, frame=first_frame)
 
     @classmethod
     def get_expected_resolution(cls, instance):
         data = instance.data["assetEntity"]["data"]
         return data["resolutionWidth"], data["resolutionHeight"]
+
+    @classmethod
+    def get_tool_resolution(cls, tool, frame):
+        """Return the 2D input resolution to a Fusion tool
+
+        If the current tool hasn't been rendered its input resolution
+        hasn't been saved. To combat this, add an expression in
+        the comments field to read the resolution
+
+        Args
+            tool (Fusion Tool): The tool to query input resolution
+            frame (int): The frame to query the resolution on.
+
+        Returns:
+            tuple: width, height as 2-tuple of integers
+
+        """
+        if tool["Comments"][frame] is None:
+            raise PublishValidationError(
+                "Cannot get resolution info for frame '{}'.\n\n "
+                "Please check that saver has connected input.".format(
+                    frame
+                )
+            )
+
+        comp = tool.Composition
+
+        # False undo removes the undo-stack from the undo list
+        with comp_lock_and_undo_chunk(comp, "Read resolution", False):
+            # Save old comment
+            old_comment = ""
+            has_expression = False
+
+            if tool["Comments"][frame] != "":
+                if tool["Comments"].GetExpression() is not None:
+                    has_expression = True
+                    old_comment = tool["Comments"].GetExpression()
+                    tool["Comments"].SetExpression(None)
+                else:
+                    old_comment = tool["Comments"][frame]
+                    tool["Comments"][frame] = ""
+
+            # Get input width
+            tool["Comments"].SetExpression("self.Input.OriginalWidth")
+            width = int(tool["Comments"][frame])
+
+            # Get input height
+            tool["Comments"].SetExpression("self.Input.OriginalHeight")
+            height = int(tool["Comments"][frame])
+
+            # Reset old comment
+            tool["Comments"].SetExpression(None)
+            if has_expression:
+                tool["Comments"].SetExpression(old_comment)
+            else:
+                tool["Comments"][frame] = old_comment
+
+            return width, height
```

Comment:
Having this logic outside of the class as a function makes the function **way** easier to test with - since you can call it separate from the class and can easily test it in isolation. As such, I'd personally argue this change moving it here makes it harder to maintain.

I prefer myself to debug these types of things in isolation to ensure whatever code is written doesn't get influenced by anything else. It also allows me to test the behavior without needing to actually perform a publish, meaning I can run this easily a few times within a few seconds instead of needing to go through the publishing process which will easily take a few seconds each.

Reply:
How you print stuff if the function is outside of the class? I tried 'print' and it didn't work (I think).

Reply:
I meant more that usually I call these methods manually for debugging outside of the scope of the publisher - in this instance I'd likely copy paste it to a temporary location since it's not in `lib` so maybe the point is less valid.